internal/blocksync/pool.go

```diff
@@ -130,23 +130,31 @@ func (*BlockPool) OnStop() {}
 
 // spawns requesters as needed
 func (pool *BlockPool) makeRequestersRoutine(ctx context.Context) {
-	for {
-		if !pool.IsRunning() {
-			break
-		}
-
+	for pool.IsRunning() {
 		_, numPending, lenRequesters := pool.GetStatus()
 		switch {
 		case numPending >= maxPendingRequests:
-			// sleep for a bit.
-			time.Sleep(requestIntervalMS * time.Millisecond)
+			timer := time.NewTimer(requestIntervalMS * time.Millisecond)
+			select {
+			case <-ctx.Done():
+				timer.Stop()
```

Comment:
Since we do this with the same interval in each of the loop cases, maybe it would be simpler to create a ticker once at the top level:

```go
ticker := time.NewTicker(requestInterval)
defer ticker.Stop()
// ...
case blah:
	select {
	case <-ctx.Done():
		return
	case <-ticker.C:
		pool.removeTimeoutPeers()
	}
// etc.
```

Rationale: There is a minor efficiency in not creating and reaping a goroutine every time we want a new interval, and deferring the stop ensures we do it even if something panics. That's not essential, but brings peace of mind. More importantly, unless we care about exact timing, this saves us from having to remember to add new timers if we add more cases.

(If we care about exact timing—e.g., we want to be sure we wait the whole interval each time—the same structure would work, but using `NewTimer` instead of `NewTicker` and calling `timer.Reset(requestInterval)` at the head of each case. Given the way this works, I'm pretty sure we don't care about that, though)

Also relatedly: I suggest redefining `requestIntervalMS` as:

```go
const requestInterval = 2*time.Milliseconds
```

so that we don't have to do any conversion at the usage.

Reply:
I've updated this. To me it seems like a timer's still the best option, along with `timer.Reset`, to preserve the existing behavior. I'm always wary of changing subtle timing assumptions in this codebase in case things go :boom:

The two cases in the switch statement actually seem to be executing the same code, so I've merged them into one now.

I've also simplified the `requestInterval` definition.

Reply:
I'm sorry, forgive me, this wound up being more disruptive than I realized or intended.

I implicitly assumed we didn't care about the exact interval, only that we not update the pool "too often". If we have tests that depend on the exact interval, however (even if they shouldn't), then I think we have a simpler approach.

Given that the interval is only a couple of milliseconds, I'd actually suggest we not use a timer _or_ a ticker, but go back to calling `time.Sleep`. If we were sleeping a long time (10s of seconds, say), being able to wake early would matter—but I'd argue ot's fine if we wait for a couple of miliiseconds after a context cancels to respond to it.

In which case, I think we can turn all these into:

```go
case blah:
   time.Sleep(requestInterval)
   if ctx.Err() != nil {
      return
   }
   pool.doTheAppropriateThing()
```

That eliminates corner cases with draining and resetting the channel.

Reply:
I've never polled a context's `Err()` method before. That shouldn't cause any problems, should it? It is thread-safe, right?

I've updated the code now to use this approach.

Reply:
It is safe (promised as such by the documentation). I keep forgetting it's an option because in the early days of the context package it specifically _wasn't_ promised, and I still forget sometimes. 😀